### PR TITLE
[RLlib] Fix TorchRLModule.self.training always True during evaluation

### LINKED ISSUE #61961: [RLlib] RLModule self.training is set to True even during evaluation
### What happened + What you expected to happen

1. When I run `algo.evaluate()`, if I log `print(self.training)` inside my RLModule's forward function, it prints `True`. This makes me think that `model.eval()` was never called.
2. When you are running evaluation, shouldn't `self.training` be set to `False` (which implies that `model.eval()` was called).

### Versions / Dependencies

ray==2.43.0
torch==2.6.0

MacOS 15.7.3
Python 3.11.5

### Reproduction script

I create the following algorithm:
```
agentConfig = (
        PPOConfig()
        .environment(env)
        .api_stack(
            enable_rl_module_and_learner=True, enable_env_runner_and_connector_v2=True
        )
        .rl_module(
            rl_module_spec=RLModuleSpec(
                module_class=module,
                model_config={
                    "hiddenSize": args.hiddenSize,
                    "numLayers": args.numLayers,
                    "inputSize": visionRange * 2 + 1 if args.fogged else mazeSize,
                    "max_seq_len": args.memoryLen,
                    "mazeSize": mazeSize,
                },
            ),
        )
        .learners(num_gpus_per_learner=1 if torch.cuda.is_available() else 0)
        .evaluation(
            evaluation_num_env_runners=1 if args.debug else 8,
            evaluation_duration_unit="episodes",
            evaluation_duration=1 if args.debug else 128,
        )
        .training(
            lr=args.lr,
            entropy_coeff=0.01,
            learner_class=PPOTorchLearnerWithSelfPredLoss,
            learner_config_dict=config,
        )
    )
agent = agentConfig.build_algo()
agent.evaluate()
```

### Issue Severity

Low

### PR DESCRIPTION
**Pull request by gambletan**
## Why are these changes needed?

Fixes #61961.

`algo.evaluate()` never called `model.eval()` on the `RLModule`, so
`self.training` was always `True` even during evaluation. This caused:

- `BatchNorm` running-statistics to be updated during evaluation rollouts (incorrect)
- `Dropout` layers remaining active during evaluation (incorrect)
- User code inside `_forward_inference` / `_forward_exploration` unable to
  distinguish between training and inference contexts via `self.training`

The old API stack (`TorchPolicyV2`) already handled this correctly by calling
`self.model.eval()` in `_compute_action_helper`. The new RLModule API stack
was missing the equivalent.

## What changes were made?

`TorchRLModule` now overrides the three public forward methods to bracket each
call with the correct PyTorch training-mode flag:

| Method | Training mode inside call | Restored to |
|---|---|---|
| `forward_inference` | `eval` (`self.training == False`) | `train` |
| `forward_exploration` | `eval` (`self.training == False`) | `train` |
| `forward_train` | `train` (`self.training == True`) | — |

A `try/finally` guard in `forward_inference` and `forward_exploration` ensures
`train` mode is restored even when the forward pass raises an exception.

The `_forward_inference` / `_forward_exploration` private overrides (which
users are encouraged to customise) are unchanged — they continue to wrap the
call in `torch.no_grad()` for performance.

## Tests

Added `TestRLModule.test_training_mode_during_forward_passes` in
`rllib/core/rl_module/torch/tests/test_torch_rl_module.py`.

The test subclasses `VPGTorchRLModule`, captures `self.training` inside each
`_forward_*` hook, and asserts:

- `self.training == False` during `_forward_inference`
- `self.training == False` during `_forward_exploration`
- `self.training == True` during `_forward_train`
- Module is back in `train` mode after all forward passes complete

## Checklist

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR
- [x] I've run `scripts/format.sh` to lint the changes in this PR
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/
- [x] I've added any new APIs to the API stability tags.  For fixes to broken tests, skip this step.
- [x] I've made sure existing tests are not broken.
- [x] Testing Strategy
  - [x] Unit tests
  - [ ] Release tests
  - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by gambletan** [comment]
Will rebase with `--signoff` and resolve the merge conflict shortly.

File: rllib/core/rl_module/torch/torch_rl_module.py [hunk 1]
```diff
@@ -83,26 +83,46 @@ def compile(self, compile_config: TorchCompileConfig):
...
+        delegating to the parent implementation and restores ``train`` mode
+        afterwards.  This ensures that layers such as ``BatchNorm`` and
+        ``Dropout`` behave correctly during evaluation and that
+        ``self.training`` reads ``False`` inside any user-defined
+        ``_forward_inference`` override.
+        """
         self.eval()
         try:
-            return self._forward_inference(batch, **kwargs)
+            return super().forward_inference(batch, **kwargs)
```
Comment (pseudo-rnd-thoughts) [inline rllib/core/rl_module/torch/torch_rl_module.py:97]: This isn't equivalent, `self._forward_inference` has a `no_grad` while the super version is missing this. 
https://github.com/ray-project/ray/blob/275f7ea1a35b1a65c7f1b9c9be59ea2c16d214dc/rllib/core/rl_module/torch/torch_rl_module.py#L109-L114

https://github.com/ray-project/ray/blob/275f7ea1a35b1a65c7f1b9c9be59ea2c16d214dc/rllib/core/rl_module/rl_module.py#L597-L609

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]
Thanks for the PR and spotting that @gambletan. 
We should fix it but using the super version isn't equivalent to the private version. 
I'm a tad confused why we need to change `self._forward_inference` to `super().forward_inference`, could you explain?

File: rllib/core/rl_module/torch/torch_rl_module.py [hunk 2]
```diff
@@ -81,6 +81,51 @@ def compile(self, compile_config: TorchCompileConfig):
...
+        """Forward-pass for loss computation with train mode explicitly set.
+
+        Ensures ``self.training == True`` is set before the forward pass so
+        that the behaviour is well-defined even when the module was previously
+        used for inference on the same process (e.g. a local EnvRunner that
+        also runs learning updates).
+        """
+        self.train()
+        return super().forward_train(batch, **kwargs)
+
```
[the remark this answers is not in the document]
Comment (pseudo-rnd-thoughts) [reply rllib/core/rl_module/torch/torch_rl_module.py:128]: This might be helpful to make explicit whats happening
